fdbrpc/TokenSign.cpp

```diff
@@ -50,117 +64,529 @@ namespace {
+			auto typValue = StringRef(reinterpret_cast<const uint8_t*>(typ.GetString()), typ.GetStringLength());
+			if (typValue != "JWT"_sr)
+				return false;
+			return true;
 		}
+	}
+	return false;
+}
+
+template <class FieldType>
+bool parseField(Arena& arena, Optional<FieldType>& out, const rapidjson::Document& d, const char* fieldName) {
+	if (!d.HasMember(fieldName))
+		return true;
+	auto const& field = d[fieldName];
+	static_assert(std::is_same_v<StringRef, FieldType> || std::is_same_v<FieldType, uint64_t> ||
+	              std::is_same_v<FieldType, VectorRef<StringRef>>);
+	if constexpr (std::is_same_v<FieldType, StringRef>) {
+		if (!field.IsString())
+			return false;
+		out = StringRef(arena, reinterpret_cast<const uint8_t*>(field.GetString()), field.GetStringLength());
+	} else if constexpr (std::is_same_v<FieldType, uint64_t>) {
+		if (!field.IsUint64())
+			return false;
+		out = field.GetUint64();
+	} else {
+		if (!field.IsArray())
+			return false;
+		if (field.Size() > 0) {
+			auto vector = new (arena) StringRef[field.Size()];
+			for (auto i = 0; i < field.Size(); i++) {
+				if (!field[i].IsString())
+					return false;
+				vector[i] = StringRef(
+				    arena, reinterpret_cast<const uint8_t*>(field[i].GetString()), field[i].GetStringLength());
+			}
+			out = VectorRef<StringRef>(vector, field.Size());
+		} else {
+			out = VectorRef<StringRef>();
+		}
+	}
+	return true;
+}
+
+bool parsePayloadPart(Arena& arena, TokenRef& token, StringRef b64urlPayload) {
+	auto tmpArena = Arena();
+	auto [payload, valid] = base64url::decode(tmpArena, b64urlPayload);
+	auto d = rapidjson::Document();
+	d.Parse(reinterpret_cast<const char*>(payload.begin()), payload.size());
+	if (d.HasParseError()) {
+		TraceEvent(SevWarnAlways, "TokenPayloadJsonParseError")
+		    .suppressFor(10)
+		    .detail("Payload", payload.toString())
+		    .detail("Message", GetParseError_En(d.GetParseError()))
+		    .detail("Offset", d.GetErrorOffset());
 		return false;
 	}
+	if (!d.IsObject())
+		return false;
+	if (!parseField(arena, token.issuer, d, "iss"))
+		return false;
+	if (!parseField(arena, token.subject, d, "sub"))
+		return false;
+	if (!parseField(arena, token.audience, d, "aud"))
+		return false;
+	if (!parseField(arena, token.tokenId, d, "jti"))
+		return false;
+	if (!parseField(arena, token.issuedAtUnixTime, d, "iat"))
+		return false;
+	if (!parseField(arena, token.expiresAtUnixTime, d, "exp"))
+		return false;
+	if (!parseField(arena, token.notBeforeUnixTime, d, "nbf"))
+		return false;
+	if (!parseField(arena, token.keyId, d, "kid"))
+		return false;
+	if (!parseField(arena, token.tenants, d, "tenants"))
+		return false;
+	return true;
+}
+
+bool parseSignaturePart(Arena& arena, TokenRef& token, StringRef b64urlSignature) {
+	auto [sig, valid] = base64url::decode(arena, b64urlSignature);
+	if (valid)
+		token.signature = sig;
+	return valid;
+}
+
+bool parseToken(Arena& arena, TokenRef& token, StringRef signedToken) {
```

Comment:
It's tricky because exceptions are more expensive than returns, and `Error` doesn't allow attachments such as string messages. Maybe I'll take an optional `const char**` or `string_view*` argument to forward error messages if not null.